src/cascadia/TerminalControl/TermControlUiaProvider.cpp

```diff
@@ -82,49 +87,73 @@ IFACEMETHODIMP TermControlUiaProvider::get_FragmentRoot(_COM_Outptr_result_maybe
 
 const COORD TermControlUiaProvider::GetFontSize() const
 {
-    return _termControl.GetActualFont().GetSize();
+    return _termControl->GetActualFont().GetSize();
 }
 
 const winrt::Windows::UI::Xaml::Thickness TermControlUiaProvider::GetPadding() const
 {
-    return _termControl.GetPadding();
+    return _termControl->GetPadding();
 }
 
-std::deque<UiaTextRangeBase*> TermControlUiaProvider::GetSelectionRanges(_In_ IRawElementProviderSimple* const pProvider)
+HRESULT TermControlUiaProvider::GetSelectionRanges(_In_ IRawElementProviderSimple* pProvider, _Out_ std::deque<ComPtr<UiaTextRangeBase>>& result)
+try
 {
-    std::deque<UiaTextRangeBase*> result;
+    typename std::remove_reference<decltype(result)>::type temporaryResult;
+
+    std::deque<ComPtr<UiaTextRange>> ranges;
+    RETURN_IF_FAILED(UiaTextRange::GetSelectionRanges(_pData, pProvider, ranges));
 
-    auto ranges = UiaTextRange::GetSelectionRanges(_pData, pProvider);
     while (!ranges.empty())
     {
-        result.emplace_back(ranges.back());
+        temporaryResult.emplace_back(std::move(ranges.back()));
         ranges.pop_back();
     }
 
-    return result;
+    std::swap(result, temporaryResult);
+    return S_OK;
 }
+CATCH_RETURN();
 
-UiaTextRangeBase* TermControlUiaProvider::CreateTextRange(_In_ IRawElementProviderSimple* const pProvider)
+HRESULT TermControlUiaProvider::CreateTextRange(_In_ IRawElementProviderSimple* const pProvider, _COM_Outptr_result_maybenull_ UiaTextRangeBase** ppUtr)
 {
-    return UiaTextRange::Create(_pData, pProvider);
+    RETURN_HR_IF_NULL(E_INVALIDARG, ppUtr);

```

Comment:
After checking that the pointer is valid (this line), you should set the value to `nullptr` to fill it up with the null value in case the `RETURN_IF_FAILED` is hit and an early return occurs. It's part of the Out SAL contract to always fill the result with *something* on all codepaths (presuming its a valid target location)

Reply:
You will need to check and apply this pattern a bunch... I see a lot of instances of this.

Reply:
In many cases... it doesn't look like you need to be using the temporary local in these functions at all. You could just be
1. check that the outpointer is valid
2. fill outpointer with null
3. call child function return if failed to fill up the outpointer
4. return S_OK

Reply:
We need the temp var because `MakeAndInitialize` can't create a `UiaTextRangeBase`. Other than that, applied the feedback here.